src/lib.rs

```diff
@@ -1,110 +1,451 @@
+//! # use roff::*;
+//! let doc = Builder::default().text(vec![roman("hello, world")]).build().render();
+//! assert!(doc.ends_with("hello, world\n"));
+//! ```
+//!
+//! [ROFF]: https://en.wikipedia.org/wiki/Roff_(software)
+//! [groff(7)]: https://manpages.debian.org/bullseye/groff/groff.7.en.html
 
-#[derive(PartialEq, Eq)]
+#![deny(missing_docs)]
+
+use std::io::Write;
+use std::write;
+
+// A preamble added to the start of rendered output.
+//
+// This defines a string variable that contains an apostrophe. For
+// historical reasons, there seems to be no other portable way to
+// represent apostrophes across various implementations of the ROFF
+// language. In implementations that produce output like PostScript
+// or PDF, an apostrophe gets typeset as a right single quote, which
+// looks different from an apostrophe. For terminal output ("ASCII"),
+// such as when using nroff, an apostrophe looks indistinguishable
+// from a right single quote. For manual pages, and similar content,
+// an apostrophe is more generally desired than the right single
+// quote, so we convert all apostrophe characters in input text into
+// a use of the string variable defined in the preamble.
+//
+// The special handling of apostrophes is avoided in the
+// [`to_roff`](Roff::to_roff) method, but it's used in the
+// [`render`](Roff::render) and [`to_writer`](Roff::to_writer)
+// methods.
+//
+// See: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=507673#65
+const APOSTROPHE_PREABMLE: &str = r#".ie \n(.g .ds Aq \(aq
+.el .ds Aq '
+"#;
+
+// Use the apostrophe string variable.
+const APOSTROPHE: &str = r"\*(Aq";
+
+#[derive(Eq, PartialEq)]
+enum Apostrophes {
+    Handle,
+    DontHandle,
+}
+
+/// A ROFF document, consisting of lines.
+///
+/// Lines are either control lines (requests that are built in, or
+/// invocations of macros), or text lines.
+///
+/// # Example
+///
+/// ```
+/// # use roff::*;
+/// let doc = Builder::default()
+///     .control("TH", ["FOO", "1"])
+///     .control("SH", ["NAME"])
+///     .text([roman("foo - do a foo thing")])
+///     .build()
+///     .render();
+/// assert!(doc.ends_with(".TH FOO 1\n.SH NAME\nfoo \\- do a foo thing\n"));
+/// ```
+#[derive(Debug, PartialEq, Eq, Default)]
 pub struct Roff {
-    title: String,
-    section: i8,
-    content: Vec<Section>,
+    lines: Vec<Line>,
 }
 
 impl Roff {
-    pub fn new(title: &str, section: i8) -> Self {
-        Roff {
-            title: title.into(),
-            section,
-            content: Vec::new(),
-        }
+    /// Append a control line.
+    ///
+    /// The line consist of the name of a built-in command or macro,
+    /// and some number of arguments. Arguments that contain spaces
+    /// will be enclosed on double quotation marks.
+    pub fn control<'a>(&mut self, name: &'a str, args: impl Into<Vec<&'a str>>) {
+        self.lines.push(Line::control(
+            name.to_string(),
+            args.into().iter().map(|s| s.to_string()).collect(),
```

Comment:
Oh, if we're doing an `iter` here, then its probably better to accept an `impl IntoIterator<Item=&'a str>` than `impl Into<Vec<_>>`.